### PR TITLE
Snapshotting design doc

### PR DESCRIPTION
**Pull request by ciwang**
**What this PR does / why we need it**:

Design doc for volume snapshotting feature (alpha)

**Special notes for your reviewer**:

**Release note**:

``` release-note
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31919)

### CODE REVIEW

File: docs/design/volume-snapshotting.md [hunk 1]
```diff
@@ -0,0 +1,521 @@
+Kubernetes Snapshotting Proposal
+
+**Status:** Draft (as of 4 August 2016)
+
+**Authors: **[Cindy Wang](mailto:ciwang@google.com)
+
+[[TOC]]
```
Comment (saad-ali) [inline docs/design/volume-snapshotting.md]: Remove

File: docs/design/volume-snapshotting.md [hunk 2]
```diff
@@ -0,0 +1,521 @@
+Kubernetes Snapshotting Proposal
```
Comment (saad-ali) [inline docs/design/volume-snapshotting.md]: Make title a top level heading.

File: docs/design/volume-snapshotting.md [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+
+**Status:** Draft (as of 4 August 2016)
+
+**Authors: **[Cindy Wang](mailto:ciwang@google.com)
```
Comment (saad-ali) [inline docs/design/volume-snapshotting.md]: Use your GitHub profile instead of your email

File: docs/design/volume-snapshotting.md [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+
+    * Case: Failure within Kubernetes, such as controller restarts
+
+        *  If the master restarts in the middle of a snapshot operation, then the controller does not know whether or not the operation succeeded. However, since the annotation has not been deleted, the controller will retry, which may result in a crash loop if the first operation has not yet completed. This issue will not be addressed in the alpha version, but future versions will need to address it by persisting state.
+
+## Solution Overview
+
+Snapshot operations will be triggered by [annotations](http://kubernetes.io/docs/user-guide/annotations/) on PVC API objects.
+
+* **Create: **
```
Comment (saad-ali) [inline docs/design/volume-snapshotting.md]: Fix the formatting here.

File: docs/design/volume-snapshotting.md [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+
+    * Key: volume.alpha.kubernetes.io/snapshot-create
+
+    * Value: [snapshot name]
+
+* **List:** 
+
+    * Key: [snapshot name]
```
Comment (saad-ali) [inline docs/design/volume-snapshotting.md]: Add a prefix here.

**Comment by saad-ali** [comment]
Some minor comments. Otherwise LGTM

**Comment by saad-ali** [comment]
```
FAIL: changes needed but not made due to --verify
/go/src/k8s.io/kubernetes/docs/ is out of date. Please run hack/update-munge-docs.sh
FAILED   hack/make-rules/../../hack/verify-munge-docs.sh	5s
```

**Comment by saad-ali** [comment]
LGTM please squash commits.

**Comment by gr4701** [comment]
*[docs/design/volume-snapshotting.md, line 193 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31919#-KSId58QyWgr-magq5N4:-KSId58QyWgr-magq5N5:b-1dt4oj) ([raw file](https://github.com/kubernetes/kubernetes/blob/3a28c09efb81a2c8e836d9cc415e3ad0017728f1/docs/design/volume-snapshotting.md#L193)):*
> ```Markdown
> To create a snapshot:
> 
> * Acquire operation lock for volume so that no other attach or detach operations can be started for volume.
> ```

is this to limit only one snapshot request at a time ? if so why ? if the volume-provider can handle multiple attach/detach operations then why limit them ? and if the volume-provider is not able to support that; it can return an error or a retry response. > 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31919)*

**Comment by saad-ali** [comment]
@gr4701 That is incorrect, it should read `Acquire operation lock for volume so that no other snapshotting operations can be started for volume`. Feel free to send out a PR to correct.

**Pull request merged**